### DISCUSSION TITLE
Consulta misma JoinTable con varias columnas

### DISCUSSION THREAD

**Discussion by SebastianNLerner** [dds-jv/foro#107] [Consultas de persistencia] [answered]
Buenas,

Estaba tratando de codear una JoinTable con varias columnas de la siguiente manera

`@Entity`
`public class Partido {`
	...	
	`@ManyToMany
	@JoinTable(name = "Participacion")
	List<Jugador> jugadores = new ArrayList<Jugador>();`

`	@ManyToMany
	@JoinTable(name = "Participacion")
	List<Paleta> paletas = new ArrayList<Paleta>();
`
Lo que logre crear fue una tabla intermedia llamada "**Participación**" en la que tengo las columnas **paletas_codigo** , **jugadores_id** y **canchas_id**, todas como **PK**.

La duda que tengo es cómo podría hacer para que la columna **paletas_codigo** sea solo **FK**, y no forme parte de la PK de Participación.

Muchas gracias

**Comment by gastonprieto** [comment]
Buenas @SebastianNLerner !

La idea es que si son dos relaciones distintas no usen la misma tabla, porque te va a traer problemas ,pensa que si las 3 juntas son pk, necesitas que esten las 3, cosa no necesariamente cierto en el modelo de objetos que planteas. Porque son listas distintas. **Es importante mantener una consistencia entre estos modelos y no forzar a Hibernate a que tu modelo de datos encaje con tu modelo de objetos si no matchean conceptualmente.**

Si lo que vos queres es tener lo del ejercicio de Canchas, que tenias 1 Jugador + 1 Partido + 1 Paleta, entonces tu relación es Participación y debería estar reflejada en el modelo de objetos. Deberías tener una lista de objetos Participación y ese objeto Participación tener una relación a Jugador y una a Paleta. Y Partido tendría muchas Participaciones.

Y sobre la Pk de Participación ahora le pones un Id autogenerado o haces una clave compuesta (en general esto es mas complejo hacerlo con el ORM)

Te animas a ponerme como quedaría el modelo con estos cambios que te comento?

**Comment by SebastianNLerner** [comment]
Gracias por la respuesta Gaston!
Al final me lo respondieron en la clase del viernes, pero se entendió la idea. 
Debería hacer una clase nueva llamada Participación que contenga partido, jugador y paleta, y por otro lado, debería crear una lista de Participaciones en la clase partido. La relación entre Partido y Participación sería @OneToMany (no estoy tan seguro de esta, porque en el modelo de objetos el partido tiene una lista de participaciones, por lo cual conoce varias participaciones, pero en el relacional, el partido no debería tener ninguna lista de participaciones, y podría haber varias filas en la tabla participaciones con el mismo partido, por lo cual seria una relación ManyToOne, no se cual de las dos seria).
Las listas de jugadores y paletas en Partido se irían, y habría que agregar relaciones de ManyToOne para las paletas y los jugadores en Participación.
Creo que eso sería todo.

**Comment by gastonprieto** [comment]
Genial @SebastianNLerner, pero por las dudas te hago un par de comentairos. 

> Debería hacer una clase nueva llamada Participación que contenga partido, jugador y paleta, y por otro lado, debería crear una lista de Participaciones en la clase partido.

No necesariamente las 3, porque por ejemplo si el `Partido` tiene la lista de `Participacion`es entonces no sería necesario que la agregues en la clase `Participación`, porque la relación ya te la da el partido. Se ve eso?

>  relación entre Partido y Participación sería @OneToMany 

Exacto

> (no estoy tan seguro de esta, porque en el modelo de objetos el partido tiene una lista de participaciones, 

Correcto, esto hace que sea un `@____ToMany`

> por lo cual conoce varias participaciones, pero en el relacional, el partido no debería tener ninguna lista de participaciones, 

Si, y no solo no debería, no se puede (en algunos motores se puede, pero no es estandar :shushing_face: )

> y podría haber varias filas en la tabla participaciones con el mismo partido, 

Claro, precisamente eso es lo que vos expresas con el `@__ToMany` desde Partido, siempre pensa que la relación existe desde dos expremos, desde `Partido` (acá se ve `@OneToMany`) o desde `Participacion` acá se ve como `@ManyToOne`. Como pusiste en `Partido` la lista ya estas diciendo ambas cosas y el ORM entiende que va a tener una FK desde Participación hacia `Partido`

> por lo cual seria una relación ManyToOne, no se cual de las dos seria).

Calculo que la confusión de esto viene desde lo de arriba

**Comment by SebastianNLerner** [comment, marked as answer]
Genial Gastón, todo clarísimo, gracias por la respuesta.
